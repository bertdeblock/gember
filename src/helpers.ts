import type { EmberPackageJson } from "./types.js";

export function isV1Addon(packageJson: EmberPackageJson): boolean {
  if (isAddon(packageJson)) {
    const { version } = packageJson["ember-addon"] ?? {};

    return version === 1 || version === undefined;
  }

  return false;
}

export function isV2Addon(packageJson: EmberPackageJson): boolean {
  if (isAddon(packageJson)) {
    const { version } = packageJson["ember-addon"] ?? {};

    return version === 2;
  }

  return false;
}

function isAddon(packageJson: EmberPackageJson): boolean {
  if (Array.isArray(packageJson.keywords)) {
    return packageJson.keywords.includes("ember-addon");
  }

  return false;
}
