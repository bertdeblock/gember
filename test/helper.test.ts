import { afterEach, it } from "vitest";
import { Package } from "./helpers.ts";

let pkg: Package;

afterEach(() => pkg.cleanUp());

it("generates a function-based `.js` helper", async (ctx) => {
  pkg = await Package.create("v2-addon");

  await pkg.gember("helper", "foo");

  const content = await pkg.readFile("src/helpers/foo.js");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a class-based `.js` helper", async (ctx) => {
  pkg = await Package.create("v2-addon");

  await pkg.gember("helper", "foo", "--class");

  const content = await pkg.readFile("src/helpers/foo.js");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a function-based `.js` helper at a custom path", async (ctx) => {
  pkg = await Package.create("v2-addon");

  await pkg.gember("helper", "foo", "--path=src/-private");

  const content = await pkg.readFile("src/-private/foo.js");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a function-based `.ts` helper", async (ctx) => {
  pkg = await Package.create("v2-addon");

  await pkg.gember("helper", "foo", "--ts");

  const content = await pkg.readFile("src/helpers/foo.ts");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a class-based `.ts` helper", async (ctx) => {
  pkg = await Package.create("v2-addon");

  await pkg.gember("helper", "foo", "--class", "--ts");

  const content = await pkg.readFile("src/helpers/foo.ts");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a function-based `.ts` helper at a custom path", async (ctx) => {
  pkg = await Package.create("v2-addon");

  await pkg.gember("helper", "foo", "--path=src/-private", "--ts");

  const content = await pkg.readFile("src/-private/foo.ts");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a nested function-based `.js` helper", async (ctx) => {
  pkg = await Package.create("v2-addon");

  await pkg.gember("helper", "foo/bar");

  const content = await pkg.readFile("src/helpers/foo/bar.js");

  ctx.expect(content).toMatchSnapshot();
});
