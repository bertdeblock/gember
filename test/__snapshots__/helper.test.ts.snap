// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generates a class-based \`.js\` helper 1`] = `
"import Helper from "@ember/component/helper";

export default class foo extends Helper {
  compute(positional, named) {
    return positional;
  }
}
"
`;

exports[`generates a class-based \`.ts\` helper 1`] = `
"import Helper from "@ember/component/helper";

type Named = {};
type Positional = [];
type Return = Positional;

export interface FooSignature {
  Args: {
    Named: Named;
    Positional: Positional;
  };
  Return: Return;
}

export default class foo extends Helper<FooSignature> {
  compute(positional: Positional, named: Named): Return {
    return positional;
  }
}
"
`;

exports[`generates a function-based \`.js\` helper 1`] = `
"export default function foo(positional, named) {
  return positional;
}
"
`;

exports[`generates a function-based \`.js\` helper at a custom path 1`] = `
"export default function foo(positional, named) {
  return positional;
}
"
`;

exports[`generates a function-based \`.ts\` helper 1`] = `
"export default function foo(positional, named) {
  return positional;
}
"
`;

exports[`generates a function-based \`.ts\` helper at a custom path 1`] = `
"export default function foo(positional, named) {
  return positional;
}
"
`;

exports[`generates a nested function-based \`.js\` helper 1`] = `
"export default function fooBar(positional, named) {
  return positional;
}
"
`;
